### ISSUE TITLE
Load optimizer status from checkpoint file.

### ISSUE THREAD

**Issue by irasit** [PyTorchLightning/pytorch-lightning#12280]
## 🚀 Feature
In incremental training, we need to load optimizer status along with weights, and send to trainer to train it. 
But seems the optimizer is missing after load module from checkpoint file. 

```
ckpt_path = checkpoint_callback.best_model_path
ckpt_model = MyModule.load_from_checkpoint(ckpt_path)

ckpt_model.optimziers() # it is be empty.
trainer.fit(ckpt_model) # it will start a fresh new optimizer instead of reusing the old optimizer status from checkpoint.
```

### Motivation

In our pipeline, we load the checkpoint in CPU side (a spark cluster), and send model to GPU side (a Ray cluster) to do the remote training job. 
It works with Keras because when I load a compiled kearas model, the optimizer weight is part of it. So I can send this compiled model to GPU side to incrementally train it. 

### Pitch

When load the lightning module from checkpoint, optimizer status is a part of it, so I can pass the module to trainer to continue train it. 

```
# load model checkpoint in CPU side
ckpt_path = checkpoint_callback.best_model_path
ckpt_model = MyModule.load_from_checkpoint(ckpt_path, include_optimizer=True)
ckpt_model.optimziers() # it should show a loaded optimizer from checkpoing

# send to GPU cluster for training
# ckpt_model = deserialize(serialize(ckpt_model))

# start train job in GPU cluster.
trainer.fit(ckpt_model)  # it train with old optimizer status from checkpoint.
```

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @awaelchli @ananthsub @ninginthecloud @rohitgr7

**Comment by ananthsub** [comment]
The trainer will load the optimizer states from the checkpoint if you pass the checkpoint path as an argument to trainer.fit .

Does that work for you? @jjenniferdai

**Comment by jjenniferdai** [comment]
documentation for restoring training state instead of just model weights: https://pytorch-lightning.readthedocs.io/en/latest/common/checkpointing.html#restoring-training-state

I'll highlight/make this more explicit in upcoming docs updates

**Comment by irasit** [comment]
@ananthsub @jjenniferdai , thank you for the suggestion.  I tried 
```
# automatically restores model, epoch, step, LR schedulers, apex, etc...
trainer.fit(model, ckpt_path="some/path/to/my_checkpoint.ckpt")
```
It dose not work for my case, seems everything is restored. This is more like resume a paused training job instead of incremental training where I basically want to re-train(fine tune) old model with fresh new data. 
E.G. train base model with past 3 months data, then incrementally train the model weights with fresh 1 week data every weekend. 

In the new job, I only want to load last model weights and optimizer status from checkpoint, but use a different smaller learning rate, and starts the epoch/step from 0 with the new data.

Also another reason is with current design, I have to load model weights and optimizer checkpoint twice at two different place like.
 ```
ckpt_model = MyModule.load_from_checkpoint(ckpt_path) # in CPU cluster

trainer.fit(ckpt_model, ckpt_path=ckpt_path) # in GPU cluster
```

Instead, it is better to load the model weights and optimizer weights together in model loading step (Like how Keras load compiled model) in CPU cluster, then pass them together to GPU.

**Comment by rohitgr7** [comment]
so in short you want to resume your training from a checkpoint without loading the lr scheduler state?

If that's so, I think we should enable this somehow.

**Comment by shenoynikhil98** [comment]
@rohitgr7 Any update on this? Wanted to use it for a similar use case (compression fine-tuning).

**Comment by lminer** [comment]
Could we reopen this? I'm looking for the same functionality.

**Comment by MaximeB3N** [comment]
I'm also interested in this feature. At the end of a fine-tune, when I resume training, it causes a huge drop in performance that takes more than 20 hours to restore. I've tried to implement this on my own but haven't found a solution yet. The only temporary solution I found was to use a learning rate warm-up, to avoid huge drops when training resumes.

**Comment by taesungp** [comment]
+1 on reopening this.

**Comment by MaximeB3N** [comment]
I finally found a way to load the optimizer states from the checkpoint. Below, I provide an example of code to load it into the optimizer when the argument `cktp_path` was not provided inside the `trainer.fit(...)`.


```python

class PlModel(pl.LightningModule):
    def __init__(self, hparams):
        super().__init__()
        self.hparams = hparams

        self.model = self._generate_model()

        self.load_checkpoint_mode = self.hparams.load_checkpoint_mode # soft, hard, mixed
        # soft: load only the model weights
        # hard: load the model weights and the same optimizer with its optimizer state (to resume a training), but the optimizer type must be the same
        # mixed: load the model weights and the optimizer states (its hyperparameters can be overwritten), but the optimizer type must be the same

    ... 

    def _merge_optimizer_state_dict(self, optimizer, checkpoint):
        
        optimizer_states = checkpoint['optimizer_states']

        if len(optimizer_states) > 1:
            UserWarning("We use more than one optimizer, the loading is not implemented for this case")

        checkpoint_state_dict = optimizer_states[0]
        flag_state_dict = optimizer.state_dict()

        flag_state_dict['state'] = checkpoint_state_dict['state']

        if self.load_checkpoint_mode == 'mixed':
            # update the learning of the checkpoint with the learning of the flag
            # to mimic the original behavior, otherwise throws: 
            # "loaded state dict contains a parameter group that doesn't match the size of optimizer's group"
            flag_state_dict["param_groups"][0]["params"] = [0] # might differ in your case
            optimizer.load_state_dict(flag_state_dict)
        
        elif self.load_checkpoint_mode == 'hard':
            
            optimizer.load_state_dict(checkpoint_state_dict)

        
        return optimizer

    def configure_optimizers(self):
        # to generate your own optimizer,
        # but taking into account the mode
        # in mode "hard" or "mixed" you must pass the optimizer type

        optim = self._generate_optim()

        if self.load_checkpoint_mode != 'soft':
            checkpoint = torch.load(self.hparams.pl_checkpoint)

            checkpoint_optim_type = checkpoint['hyper_parameters']['optim_type']
            freeze_ratio = checkpoint['hyper_parameters']['freeze_ratio']
            
            # former freeze ratio vs current freeze ratio
            if freeze_ratio != self.freeze_ratio:
                print("Warning: the freeze ratio of the checkpoint is different from the freeze ratio of the flag, the soft behavior is done by default")
                print(f"Checkpoint freeze ratio: {freeze_ratio}")
                print(f"Flag freeze ratio: {self.hparams.freeze_ratio}")


            elif checkpoint_optim_type == self.hparams.optim_type:
                print("Loading optimizer...")
                optimizer = self._merge_optimizer_state_dict(optimizer, checkpoint)
                print('Optimizer loaded')

        # to generate your own scheduler
        scheduler = self._generate_scheduler(optim)

        return {
            "optimizer": optim,
            "lr_scheduler": {
                "scheduler": scheduler,
                "interval": self.hparams.schedule_unit
            }
        }
```
However, this works only when we use the same freeze ratio and optimizer type. In those case, I didn't find any solution but it covers most of the practical cases.

**Comment by albertfgu** [comment]
:+1: to this. It seems like a reasonable use case to want to restore the full LightningModule checkpoint from a `.ckpt` (including optimizer states) without actually resuming training. E.g. if one just wants to inspect the optimizer states at the time of the checkpoint.

In other words, there should be a way to restore the full model that's decoupled from `trainer.fit`

**Comment by Zyriix** [comment]
A simple way is loading the checkpoint, and replacing the things you want.

Like if I want to replace the hyperparameters, I can do this:

checkpoint = torch.load(ckpt_path)
checkpoint['hyper_parameters'] = model.parameters
torch.save(checkpoint, tmp_ckpt_path)
trainer.fit(model,tmp_ckpt_path)

I believe other things can be done in the similar way.
